test/test_type_promotion.py

```diff
@@ -430,6 +431,104 @@ def test_transpose(self, device):
         a = torch.tensor([[True, True], [False, True]], device=device)
         self.assertEqual(a.t() == 0, a.t() == False)  # noqa: E712
 
+    def _test_sparse_op_input_tensors(self, device, dtype, coalesced, zeros=True):
+        t = self._get_test_tensor(device, dtype, not zeros)
+        if coalesced:
+            s = t.to_sparse()
+        else:
+            s = t.to_sparse()
+            indices = torch.cat((s.indices(), s.indices()), 1)
+            values = torch.cat((s.values(), s.values()), 0)
+            s = torch.sparse_coo_tensor(indices=indices, values=values, size=s.size(), dtype=dtype, device=device)
+            t = s.to_dense()
+        self.assertEqual(s.is_coalesced(), coalesced)
+        return (t, s)
+
+    def _get_precision(self, dtype, coalesced):
```

Comment:
You might want to discuss testing in a world with type promotion with @mruberry. If the plan is for all (or most ops) to support type promotion, then the traditional @dtype and @prec decorators don't work because they assume that each operation is being run with one dtype.

Reply:
That's true, the @dtype and @prec decorators are a poor fit for type promotion tests. Oh well.

Guarding sparse tests behind onlyOnCPUAndCUDA is a good idea, though.